File: backend-ai/app/telemetry/otel.py
"""
OpenTelemetry → Grafana Cloud OTLP (traces + metrics).
Enable with OTEL_EXPORTER_OTLP_ENDPOINT (see backend-ai/.env.example).
"""
from __future__ import annotations

import base64
import logging
import os
from typing import Optional

logger = logging.getLogger(__name__)

_configured = False


def _trim_env(value: Optional[str]) -> str:
    if not value:
        return ""
    v = value.strip()
    if (v.startswith('"') and v.endswith('"')) or (v.startswith("'") and v.endswith("'")):
        return v[1:-1]
    return v


def _otlp_headers() -> Optional[str]:
    existing = _trim_env(os.getenv("OTEL_EXPORTER_OTLP_HEADERS"))
    if existing:
        return existing

    prebuilt = _trim_env(os.getenv("GRAFANA_OTLP_TOKEN"))
    if prebuilt:
        return f"Authorization=Basic {prebuilt}"

    instance_id = _trim_env(os.getenv("GRAFANA_INSTANCE_ID"))
    api_key = _trim_env(
        os.getenv("GRAFANA_CLOUD_API_KEY") or os.getenv("GRAFANA_API_KEY")
    )
    if instance_id and api_key:
        token = base64.b64encode(f"{instance_id}:{api_key}".encode()).decode()
        return f"Authorization=Basic {token}"

    return None


def configure() -> bool:
    """Initialize OTLP trace + metric providers. Returns True when active."""
    global _configured

    if _configured:
        return True

    if _trim_env(os.getenv("OTEL_SDK_DISABLED")).lower() == "true":
        return False

    endpoint = _trim_env(os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT"))
    if not endpoint:
        return False

    if ".vpce.grafana.net" in endpoint:
        print(
            "[otel] OTEL_EXPORTER_OTLP_ENDPOINT uses AWS PrivateLink; "
            "use the public OTLP gateway URL on Vultr.",
            flush=True,
        )

    headers_raw = _otlp_headers()
    if not headers_raw:
        print(
            "[otel] OTEL_EXPORTER_OTLP_ENDPOINT is set but auth is missing. "
            "Set GRAFANA_OTLP_TOKEN or GRAFANA_INSTANCE_ID + GRAFANA_CLOUD_API_KEY.",
            flush=True,
        )
        return False

    os.environ.setdefault("OTEL_EXPORTER_OTLP_ENDPOINT", endpoint)
    os.environ.setdefault("OTEL_EXPORTER_OTLP_HEADERS", headers_raw)

    service_name = _trim_env(os.getenv("OTEL_SERVICE_NAME")) or "engr-backend-ai"
    environment = _trim_env(os.getenv("ENVIRONMENT")) or "development"

    from opentelemetry import metrics
    from opentelemetry.exporter.otlp.proto.http.metric_exporter import (
        OTLPMetricExporter,
    )
    from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
        OTLPSpanExporter,
    )
    from opentelemetry.sdk.metrics import MeterProvider
    from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
    from opentelemetry.sdk.resources import Resource
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor

    resource = Resource.create(
        {
            "service.name": service_name,
            "deployment.environment": environment,
        }
    )

    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
    from opentelemetry import trace

    trace.set_tracer_provider(tracer_provider)

    metric_reader = PeriodicExportingMetricReader(
        OTLPMetricExporter(),
        export_interval_millis=int(
            os.getenv("OTEL_METRIC_EXPORT_INTERVAL_MS", "60000")
        ),
    )
    meter_provider = MeterProvider(resource=resource, metric_readers=[metric_reader])
    metrics.set_meter_provider(meter_provider)

    _configured = True
    print(
        f"[otel] Exporting traces + metrics for {service_name} ({environment})",
        flush=True,
    )
    return True


def instrument_fastapi(app) -> None:
    if not _configured:
        return

    from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
    from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor

    FastAPIInstrumentor.instrument_app(app)
    HTTPXClientInstrumentor().instrument()

File: backend-nest/src/otel.ts
/**
 * OpenTelemetry → Grafana Cloud OTLP (traces + metrics, RED in Application Observability).
 * Must load before Nest/Express (imported first in main.ts).
 *
 * Enable by setting OTEL_EXPORTER_OTLP_ENDPOINT on the server (see backend-nest/.env.example).
 */
import { config as loadEnv } from 'dotenv';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import {
  ATTR_DEPLOYMENT_ENVIRONMENT_NAME,
  ATTR_SERVICE_NAME,
} from '@opentelemetry/semantic-conventions';

// OTel runs before Nest ConfigModule; load backend-nest/.env here.
loadEnv({ path: ['.env', '.env.local'] });

function trimEnv(value: string | undefined): string {
  const v = value?.trim() ?? '';
  if (
    (v.startsWith('"') && v.endsWith('"')) ||
    (v.startsWith("'") && v.endsWith("'"))
  ) {
    return v.slice(1, -1);
  }
  return v;
}

function isOtelEnabled(): boolean {
  if (process.env.OTEL_SDK_DISABLED === 'true') return false;
  return Boolean(trimEnv(process.env.OTEL_EXPORTER_OTLP_ENDPOINT));
}

function warnIfPrivateLinkEndpoint(endpoint: string): void {
  if (!endpoint.includes('.vpce.grafana.net')) return;
  console.warn(
    '[otel] OTEL_EXPORTER_OTLP_ENDPOINT uses AWS PrivateLink (.vpce.grafana.net). ' +
      'That hostname only works from resources inside the linked AWS VPC. ' +
      'On Vultr or your laptop, use the public OTLP URL from Grafana Cloud → OpenTelemetry ' +
      '(e.g. https://otlp-gateway-prod-ap-south-0.grafana.net/otlp).',
  );
}

function applyGrafanaOtlpAuth(): void {
  if (process.env.OTEL_EXPORTER_OTLP_HEADERS?.trim()) return;

  const prebuilt = trimEnv(process.env.GRAFANA_OTLP_TOKEN);
  if (prebuilt) {
    process.env.OTEL_EXPORTER_OTLP_HEADERS = `Authorization=Basic ${prebuilt}`;
    return;
  }

  const instanceId = trimEnv(process.env.GRAFANA_INSTANCE_ID);
  const apiKey = trimEnv(
    process.env.GRAFANA_CLOUD_API_KEY || process.env.GRAFANA_API_KEY,
  );
  if (instanceId && apiKey) {
    const basic = Buffer.from(`${instanceId}:${apiKey}`).toString('base64');
    process.env.OTEL_EXPORTER_OTLP_HEADERS = `Authorization=Basic ${basic}`;
    return;
  }

  console.warn(
    '[otel] OTEL_EXPORTER_OTLP_ENDPOINT is set but OTLP auth is missing. ' +
      'Set GRAFANA_OTLP_TOKEN (Base64 from Grafana OpenTelemetry page) or ' +
      'GRAFANA_INSTANCE_ID + GRAFANA_CLOUD_API_KEY.',
  );
}

if (isOtelEnabled()) {
  const endpoint = trimEnv(process.env.OTEL_EXPORTER_OTLP_ENDPOINT);
  warnIfPrivateLinkEndpoint(endpoint);
  applyGrafanaOtlpAuth();

  const serviceName =
    trimEnv(process.env.OTEL_SERVICE_NAME) || 'engr-backend-nest';
  const environment = process.env.NODE_ENV || 'development';

  const sdk = new NodeSDK({
    resource: resourceFromAttributes({
      [ATTR_SERVICE_NAME]: serviceName,
      [ATTR_DEPLOYMENT_ENVIRONMENT_NAME]: environment,
    }),
    traceExporter: new OTLPTraceExporter(),
    metricReader: new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter(),
      exportIntervalMillis: Number(
        process.env.OTEL_METRIC_EXPORT_INTERVAL_MS || 60_000,
      ),
    }),
    instrumentations: [
      getNodeAutoInstrumentations({
        // Reduce noise; HTTP metrics/traces are what we need for latency.
        '@opentelemetry/instrumentation-fs': { enabled: false },
        '@opentelemetry/instrumentation-dns': { enabled: false },
      }),
    ],
  });

  sdk.start();
  console.info(`[otel] Exporting traces + metrics for ${serviceName} (${environment})`);

  const shutdown = () => {
    void sdk.shutdown();
  };
  process.once('SIGTERM', shutdown);
  process.once('SIGINT', shutdown);
}
